### PR TITLE
cache config groups: optimize user permission checks

### PR DESCRIPTION
**Pull request by codelegend**
also removed `Config.groupsranking`, and assigned it while caching.

### CODE REVIEW

File: users.js [hunk 1]
```diff
@@ -398,6 +398,34 @@ function exportUsergroups() {
 }
 importUsergroups();
 
+function cacheGroupData() {
+	var groups = Config.groups;
+	Config.groupsranking = Object.keys(groups).reverse();
```
Comment (Slayer95) [inline users.js]: This is not spec−compliant. The order of the keys is not to be trusted.

Comment (codelegend) [inline users.js]: So, should I keep the `rank` key?

**Comment by ascriptmaster** [comment]
Pretty sure we need to keep the rankings, especially when branching inheritance paths are involved. Some servers might want the player rank to be lower than global drivers for mute purposes, while some might want it higher (like the example, which puts them on the same level as RO)

**Comment by Zarel** [comment]
I thought the idea was to delete this line wasn't it? `exports.groupsranking = [' ', '+', '%', '@', '\u2605', '#', '&', '~'];`

**Comment by Zarel** [comment]
@ascriptmaster, the rankings are still there, they're just autogenerated (currently by `Config.groupsranking` but I was hoping to do it just by `for-in` order) instead of needing to be specified explicitly

**Comment by ascriptmaster** [comment]
Yeah, but I was just thinking that some servers want the player rank to outrank global drivers, and some servers want the opposite

**Comment by codelegend** [comment]
@Zarel , my original commit was to determine ranks using `for-in` order,
but @Slayer95 said
> The order of the keys is not to be trusted, specially not in a component as critical as permissions.

**Comment by Zarel** [comment]
You can change it from an Object to an Array, then.

**Comment by codelegend** [comment]
Making it an array will require a whole lot of refactoring, coz. `Config.groups[symbol]` is used almost everywhere.

**Comment by Zarel** [comment]
Obviously `cacheGroupData` can turn it back into an Object.

**Comment by Zarel** [comment]
You could even name the array something else, to make it clearer.

**Comment by codelegend** [comment]
but that would destroy the design, ie. the `symbol: {permissions}` style. Which makes it slightly tougher to interpret or edit. Also, having `Config.groupsranking` helps to re-order groups easily without having to move the whole object declaration, and gives a quick review of the order.

**Comment by Zarel** [comment]
You just add `symbol: '~',` above `id: 'admin',`. I don't think that'd be noticeably tougher to interpret or edit, especially since it'll be changed easily.

Copy/paste can re-order groups plenty fast enough, and having groups defined in the correct order rather than a random order is nice for readability anyway.

**Comment by codelegend** [comment]
ok, all set. also documented the new format, and a few missing permissions

File: users.js [hunk 2]
```diff
@@ -398,6 +398,41 @@ function exportUsergroups() {
...
+			if (cacheGroup(groupData['inherit'], inheritGroup)) {
+				Object.merge(groupData, inheritGroup, false, false);
+			}
+			delete groupData['inherit'];
+		}
+		return (cachedGroups[sym] = true);
+	}
+
+	var numGroups = grouplist.length;
+	for (var i in grouplist) {
```
Comment (Slayer95) [inline users.js]: <code>for..in</code> here defeats the purpose of using an array :|

File: users.js [hunk 3]
```diff
@@ -398,6 +398,41 @@ function exportUsergroups() {
...
+		var groupData = grouplist[i];
+		groupData.rank = numGroups - i - 1;
```
Comment (Slayer95) [inline users.js]: Do we really need to get the complement?

File: users.js [hunk 4]
```diff
@@ -398,6 +398,41 @@ function exportUsergroups() {
 }
 importUsergroups();
 
+function cacheGroupData() {
+	var grouplist = Config.grouplist;
+	var groups = Config.groups = {};
```
Comment (Slayer95) [inline users.js]: Converting this to <code>Object.create(null)</code> would be a nice improvement.

**Comment by codelegend** [comment]
@Slayer95 , thanks for the suggestions. Updated now.
(helped me learn rebasing across commits too).

**Comment by codelegend** [comment]
@Zarel ?

**Comment by Zarel** [comment]
Looks good. Is this backwards compatible?

**Comment by codelegend** [comment]
like compatible with the old config.groups format?? no.
Anyways, why would it matter, people who update to this will automatically notice the change in that (and edit their config files)

**Comment by Zarel** [comment]
...how would they automatically notice? Would they at least get an informative error message? Is there a way to migrate without downtime?

**Comment by codelegend** [comment]
well, they cant migrate without downtime if they have `Config.watchconfig` set to `true`.
Anyways, I'll add a error message, saying that their config file is not supported.

**Comment by codelegend** [comment]
Ok, this shows a warning message if used with the old `Config.groups` format.

File: users.js [hunk 5]
```diff
@@ -398,6 +398,51 @@ function exportUsergroups() {
...
+		return (cachedGroups[sym] = true);
+	}
+
+	var numGroups = grouplist.length;
+	for (var i = 0; i < numGroups; i++) {
+		var groupData = grouplist[i];
+		groupData.rank = numGroups - i - 1;
+		groups[groupData.symbol] = groupData;
+		Config.groupsranking.unshift(groupData.symbol);
+	}
```
Comment (Zarel) [inline users.js]: Why don't you just make it backwards compatible? As far as I can tell, all you have to do is wrap lines 430-436 in `if (grouplist)` and move ` = Object.create(null)` from line 412 at the top of the new `if`-block.

Comment (Zarel) [inline users.js]: I can understand not making it backwards compatible if it's hard to do, but this seems really easy.

Comment (codelegend) [inline users.js]: ok

**Comment by codelegend** [comment]
ok. Backwards compatibility added.

**Pull request merged**